### PR TITLE
chore: better typings after transforming oas file

### PR DESCRIPTION
**Pull request by lottamus**
Happy to hop on a call to walk through the changes too!

This PR was broken off of https://github.com/stoplightio/elements/pull/1084 and is needed for https://github.com/stoplightio/elements/pull/1110.

## What's in this PR?

- Reworks how we transform OAS files into http-spec nodes to take advantage of better typings (https://github.com/stoplightio/elements/issues/867)

### CODE REVIEW

**Review by mpodlasin — APPROVED** [review]
Only minor issues in the code and it works very well, so I am approving.

File: packages/elements/src/components/API/SidebarLayout.tsx [hunk 1]
```diff
@@ -1,45 +1,59 @@
...
+      serviceNode.children.reduce((prev, current) => {
+        if (current.type === NodeType.HttpOperation) {
+          prev[current.uri] = current.data.method;
+        }
+
+        return prev;
+      }, {}),
+    [serviceNode.children],
+  );
+  const tree = React.useMemo(() => generateApiToC([serviceNode, ...serviceNode.children], serviceNode.data), [
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/API/SidebarLayout.tsx:43]: I do in the next PR, but wanted to keep this one smaller: https://github.com/stoplightio/elements/pull/1110/files#diff-46e8533247639f8e0dd666c48d0fa951feb99d6860177e4fea88ddf87ad7310bR54

File: packages/elements/src/utils/oas/types.ts [hunk 2]
```diff
@@ -13,3 +17,16 @@ export interface ISourceNodeMap {
   notMatch?: string;
   children?: ISourceNodeMap[];
 }
+
+type Node<T, D> = {
+  type: T;
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/utils/oas/types.ts:22]: I wanted to have accurate typings for combination of NodeType + data shape. For example the data object for `NodeType.HttpOperation` will always be `IHttpOperation`

https://github.com/stoplightio/elements/pull/1109/files/06ed6c42dea9bc80b80380c6b4ece523cb34947e#diff-d5f14078a2cda3f67c1f0cc7f5b4e8f67a06792be18c75f30a2e1538ff406d7dR29-R32

**Comment by philsturgeon** [comment]
Great change and thank you for giving the extra context. Looking forward to more of your PRs! 🥳

--
Phil Sturgeon
Product @ Stoplight.io

**Pull request merged**